docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,343 @@
+
+Attestation data will be appended, with any relevant base64 encoded data wrapped with the following header & footer.
+
+Example CSR:
+```
+-----BEGIN ATTESTATION DATA-----
+S25vd2luZyBtZSBBbGFuIFBhcnRyaWRnZSwga25vd2luZyB5b3UgS3ViZXJuZXRlcyBjbHVzdGVyLCBhaGEh
+-----END ATTESTATION DATA-----
+```
+The format of the attestation block is left to the provider. The Appendix section provides examples.
+
+#### Kubelet serving certificate rotation daemon
+
+In the first attempt, we will attempt to run serving certificate rotation as part of the client authenticator - it will
+check for the existence of the serving certificate files, their validity (e.g. not self signed), or request a new serving certificate. Once the new serving certificate is written, kubelet can be restarted.
+
+We may however find that it is better suited to run as a windows service manager service on the node, but its lifecycle management would be out of scope for now.
+
+There is a un untracked effort to make components automatically reload when certificates are changed, and will follow up
+as appropriate.
+
+
+#### Core Specification
+- Core Cluster API MUST provide the following implementations of CSRs and signers:
+    - `cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure` which implement an “Always Allow” type signer that provides equivalent security to Cluster API v1alpha3. This is only to be used for providers where no secure mechanism exists.
+    - `cluster.x-k8s.io/kube-apiserver-client-kubelet-tpm` and `cluster-x-k8s-io/kubelet-serving-tpm`
+        - Will implement TPM-based certificate signers and requesters based on the cloud-provider-gcp implementation.
+        - However, since the mechanism for retrieving endorsement keys varies across platforms, the TPM signer will additionally require a provider specific mechanism to provide the TPM Endorsement Key.
+
+#### Provider Specification
+
+##### All providers
+- All providers MUST insert a ProviderID within the KubernetesNodeProviderIdentifierOID extension attribute of the CSR
+- All signer names MUST be filled in by the provider’s controller in InfraCluster.Status.KubeletClientCertificateSigner and InfraCluster.Status.KubeletServingCertificateSigner if the attestation controller is running
+- All providers SHOULD implement trust-on-first-use type mechanisms to prevent replay attacks
+
+##### Insecure providers
+- An insecure provider CANNOT implement certificate rotation or kubelet serving certificate signing
+- InfraCluster.Status.KubeletClientCertificateSigner MUST be set to cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure
+- An insecure provider MUST use the cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure signer
+
+##### Secure providers
+- A secure provider MUST implement certificate rotation and kubelet server certificate signing
+- A provider must register signers of:
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-<provider>`
+    - `cluster-x-k8s-io/kubelet-serving-<provider>`
```

Comment:
Will these be easy enough to rename to omit the `x-` when the feature graduates to beta / GA / whenever?

Reply:
we should be able to graduate those with backfilling and using both values and drop `x-` ones down the road

Reply:
I'm going to defer to @vincepri for some advice here as we're using x.k8s.io elsewhere